src/Mvc/Mvc.Core/src/ModelBinding/Binders/TryParseModelBinder.cs

```diff
@@ -0,0 +1,150 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq.Expressions;
+using System.Reflection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders;
+
+/// <summary>
+/// An <see cref="IModelBinder"/> for simple types.
+/// </summary>
+internal sealed class TryParseModelBinder : IModelBinder
+{
+    private static readonly MethodInfo AddModelErrorMethod = typeof(TryParseModelBinder).GetMethod(nameof(AddModelError), BindingFlags.NonPublic | BindingFlags.Static)!;
+    private static readonly MethodInfo SuccessBindingResultMethod = typeof(ModelBindingResult).GetMethod(nameof(ModelBindingResult.Success), BindingFlags.Public | BindingFlags.Static)!;
+    private static readonly ParameterExpression BindingContextExpression = Expression.Parameter(typeof(ModelBindingContext), "bindingContext");
+    private static readonly ParameterExpression ValueProviderResultExpression = Expression.Parameter(typeof(ValueProviderResult), "valueProviderResult");
+    private static readonly MemberExpression BindingResultExpression = Expression.Property(BindingContextExpression, nameof(ModelBindingContext.Result));
+    private static readonly MemberExpression ValueExpression = Expression.Property(ValueProviderResultExpression, nameof(ValueProviderResult.FirstValue));
+    private static readonly MemberExpression CultureExpression = Expression.Property(ValueProviderResultExpression, nameof(ValueProviderResult.Culture));
+
+    private readonly Func<ValueProviderResult, ModelBindingContext, object?> _tryParseOperation;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="SimpleTypeModelBinder"/>.
+    /// </summary>
+    /// <param name="modelType">The model type.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+    public TryParseModelBinder(Type modelType, ILoggerFactory loggerFactory)
+    {
+        if (modelType == null)
```

Comment:
Honestly I would remove null checks for internal types when the inputs are devised by us. That's what the compiler's nullability enforcement is for

Reply:
I can do that, however, we decided to keep this Binder internal instead of public as the other Binders are. I wonder if, in the future, we will decide to make it public and we will miss this null check.